broker/src/bin/neon_broker.rs

```diff
@@ -0,0 +1,563 @@
+        }
+        // Note: it is important to create Subscriber before .await. If client
+        // disconnects during await, which would terminate the Future we still
+        // need to run Subscriber's drop() which will unregister it from the
+        // shared state.
+        let subscriber = Subscriber {
+            id,
+            key: sub_key,
+            sub_rx: rx,
+            registry: self.clone(),
+        };
+        // Notify existing publishers about new subscriber.
+        for pub_tx in pub_txs.iter_mut() {
+            // Closed channel is fine; it means publisher has gone.
+            pub_tx.send(announce.clone()).await.ok();
+        }
+        trace!("registered subscriber {}", id);
+        subscriber
+    }
+
+    // Unregister the subscriber
+    pub fn unregister_subscriber(&self, sub: &Subscriber) {
+        let mut ss = self.shared_state.lock().unwrap();
+        let announce_pack = ss.unregister_subscriber(sub.id, sub.key);
+        drop(ss);
+        // Notify publishers about the removal. Apart from wanting to do it
+        // outside lock, here we also spin a task as Drop impl can't be async.
+        if let Some((mut pub_txs, announce)) = announce_pack {
+            tokio::spawn(async move {
+                for pub_tx in pub_txs.iter_mut() {
+                    // Closed channel is fine; it means publisher has gone.
+                    pub_tx.send(announce.clone()).await.ok();
+                }
+            });
+        }
+        trace!("unregistered subscriber {}", sub.id);
+    }
+}
+
+// Private subscriber state.
+struct Subscriber {
+    id: SubId,
+    key: SubscriptionKey,
+    // Subscriber receives messages from publishers here.
+    sub_rx: Receiver<SafekeeperTimelineInfo>,
+    // to unregister itself from shared state in Drop
+    registry: Registry,
+}
+
+impl Drop for Subscriber {
+    fn drop(&mut self) {
+        self.registry.unregister_subscriber(self);
+    }
+}
+
+// Private publisher state
+struct Publisher {
+    id: PubId,
+    // new subscribers request to send (or stop sending) msgs them here.
+    // It could be just Receiver, but weirdly it doesn't implement futures_core Stream directly.
+    announce_rx: ReceiverStream<SubAnnounce>,
+    subs_to_all: Vec<Sender<SafekeeperTimelineInfo>>,
+    subs_to_timelines: HashMap<TenantTimelineId, Vec<SubSender>>,
+    // to unregister itself from shared state in Drop
+    registry: Registry,
+}
+
+impl Publisher {
+    // Send msg to relevant subscribers.
+    pub fn send_msg(&mut self, msg: &SafekeeperTimelineInfo) -> Result<(), Status> {
+        // send message to subscribers for everything
+        let mut cleanup_subs_to_all = false;
+        for sub in self.subs_to_all.iter() {
+            match sub.try_send(msg.clone()) {
+                Err(TrySendError::Full(_)) => {
+                    trace!("dropping message, channel is full");
+                }
```

Comment:
This Drop when full behaviour makes me wonder if per publisher [tokio::sync::broadcast](https://docs.rs/tokio/latest/tokio/sync/broadcast/index.html) channel would had been more applicable. On being full it "drops" oldest messages and resumes from oldest still reachable, compared to this which drops the later messages but continues to send older, skipping some in the middle.

Memory-wise it would allow all subscribers to follow a single queue instead of having their own, removing some cleanup lines at the cost of having more complexity inside subscriber's `try_stream!`, unsure how to handle the `subs_to_all` at least right away... EDIT: I don't think that's an issue since one subscriber only subscribes to all or specific timelines.

Reading this more I think it would help with not needing to be concerned about removal or even book keeping of subscribers.

If you already considered it but it cannot work for some reason I cannot see, could a comment be added somewhere?

Reply:
Nice, [tokio::sync::broadcast](https://docs.rs/tokio/latest/tokio/sync/broadcast/index.html) looks like a ready implementation of what a broker does, i.e. Pub/Sub. When we will rewrite the broker code, we should definitely consider using tokio::sync::broadcast.

Reply:
Cool, I didn't know it has support for dropping slow receivers messages, will try to switch.